srcts/src/components/webcomponents/layoutColumns.ts

```diff
@@ -0,0 +1,543 @@
+  static isShinyInput = false;
+
+  /**
+   * Create a new default BreakpointMap. If no column widths are specified, we
+   * auto-fit at the "sm" and "lg" breakpoints.
+   */
+  static defaultColWidths(): BreakpointMap {
+    return new Map(Object.entries({ sm: null, lg: null }));
+  }
+
+  /**
+   * The user-specified (or default) column widths, unresolved. We don't support
+   * it now, but if the number of children were to change, we can re-resolve the
+   * column width spec starting from this value.
+   */
+  colWidths!: BreakpointMap;
+
+  /**
+   * The column width specification, with auto-fit breakpoint values resolved.
+   */
+  colWidthsSpec: BreakpointColumnSpec | undefined;
+
+  /**
+   * The number of column units in a row.
+   */
+  private _colUnits = 12;
+
+  get colUnits(): number {
+    return this._colUnits;
+  }
+
+  set colUnits(val: number) {
+    this.style.setProperty("--bs-columns", `${val}`);
+    this._colUnits = val;
+  }
+
+  connectedCallback(): void {
+    // This component assumes Bootstrap's CSS grid, so we enforce the class here
+    this.classList.add("grid");
+
+    // Read and parse the column widths (the user's unresolved input)
+    this.colWidths = this._readColWidths();
+    // We don't observe or reflect the attribute, and it could be an arbitrarily
+    // big JSON string, so remove it
+    this.removeAttribute("col-widths");
+
+    // Connected callback is called before children are added to the DOM, so we
+    // need to wait for the next tick to apply the column widths.
+    setTimeout(() => {
+      this._applyColWidthsSpec();
+      this.style.removeProperty("display");
+    });
+  }
+
+  /**
+   * Reads and parses the column widths from the "col-widths" attribute.
+   * @returns A map of breakpoint to column width.
+   */
+  private _readColWidths(): BreakpointMap {
+    const attr = this.getAttribute("col-widths");
+    if (!attr) {
+      return BslibLayoutColumns.defaultColWidths();
+    }
+
+    let x = JSON.parse(attr);
+    if (Array.isArray(x)) {
+      x = { md: x };
+    }
+
+    const colWidths = new Map();
+    const breaks = ["sm", "md", "lg", "xl", "xxl"];
+
+    // Assign known bootstrap breakpoints in order
+    breaks.forEach((breakName) => {
+      if (x[breakName]) {
+        colWidths.set(breakName, isNA(x[breakName]) ? null : x[breakName]);
+        delete x[breakName];
+      }
+    });
+    // Plus any remaining breakpoints
+    Object.keys(x).forEach((breakName) => {
+      colWidths.set(breakName, isNA(x[breakName]) ? null : x[breakName]);
+    });
+
+    // fill in `md` breakpoint if only large breakpoints are specified
+    const hasLg = ["lg", "xl", "xxl"].some((size) => colWidths.has(size));
+    const hasSm = ["sm", "md"].some((size) => colWidths.has(size));
+
+    if (hasLg && !hasSm) {
+      const smallest = ["lg", "xl", "xxl"].filter((size) =>
+        colWidths.has(size)
+      );
+      colWidths.set("md", colWidths.get(smallest[0]));
+    }
```

Comment:
I think we should do this via the `.bslib-grid-item` class, e.g. using a default width of 6 column units or working out from `--bs-columns`.

Reply:
I updated this to use a fallback span of 6, following a CSS variable that's customizable on the `.bslib-grid` parent element. We tweak this as necessary when the auto-fit algorithm adjusts the `--bs-columns` variable.

Reply:
https://github.com/rstudio/bslib/pull/931/commits/4ffb9b8865eb261994c964ff5e2fabe080349227